### DISCUSSION TITLE
Spearhead Sump

### DISCUSSION THREAD

**Discussion by thorwhalen** [i2mint/poclab#1] [General]
A place to accumulate tasks, R&D, and project ideas for Spearhead Sump.

`Spearhead Sump` denotes a place where a collection of ideas for research and development (R&D) is accumulated. 

	•	Spearhead: This suggests leading, pioneering, or being at the forefront of innovation and development efforts.
	•	Sump: Traditionally, a sump is a pit or reservoir where liquids collect, but metaphorically, it can be a repository or collection point.

Combining these concepts, “Spearhead Sump” could refer to a centralized hub or repository where innovative ideas, concepts, and research initiatives are gathered and developed. It implies a dynamic, leading-edge environment dedicated to pioneering new technologies or solutions.

## Main domains

* Harnessing/Integrating AI in dev workflow and related ecosystem (blogging, presentations, etc.)
   * Code search and RAG
   * Research and publication
   * Data Prep
* i2i:
    * Diagramming
    * Extreme agility
    * Data framework
    * Language bridges (e.g. jy, ju, qh)
* Slang (mapping signals to language) and SLMs
* Music
  * Theramin
  * Sonification
  * AI accompaniment
* CI crud

**Comment by thorwhalen** [comment]
## 2024-06-17


* Notes:
  * chores (with AI)
    * documentation
    * tests
    * reviews
  * [imbed](https://github.com/thorwhalen/imbed/discussions)-related projects
    * facade the imbedding function (use [langchain](https://www.langchain.com/)?)
      * research other imbedding offerings (APIs)
      * facade them with a consistent facade
    * search-engine back and front
    * Streamlining embedding data prep
    * [RAG](https://python.langchain.com/v0.2/docs/tutorials/rag/)
  * [scraped](https://github.com/thorwhalen/scraped)
    * chrome extension (e.g. to make markdown dump, contextual search/highlighting, inject into chatGPT, etc.)
  * [jy](https://github.com/i2mint/jy) and [ju](https://github.com/i2mint/ju)
    * streamlining python widgets dev (have a look at [cosmograph](https://cosmograph.app/) widget?)
    * from python to js
      * transpiling 
      * calling py from js, through python interpreter service
    * json schema to python function (e.g. openAPI schema)
  * i2i (in collaboration with addix?)
    * [sqldol](https://github.com/i2mint/sqldol) (see [dol](https://github.com/i2mint/dol)
    * [qh](https://github.com/i2mint/qh)
    * [rjsf](https://rjsf-team.github.io/react-jsonschema-form/docs/) & fundamental UI components
    * [tabled](https://github.com/i2mint/tabled) and [doodad](https://github.com/i2mint/tabled) (smart joins and wikipedia tables)
  * github (using [hubcap](https://github.com/thorwhalen/hubcap)) info aggregator
  * [dag apps](https://github.com/i2mint/dagapp)

**Comment by sushpradi** [comment]
Jun 18 call:
- Discussed from 12:30 - 1:30 pm IST.
- Thor gave summary of above listed projects/ideas where Sushma can contribute.
- Next step:
  * Sushma to look at imbed and see how to contribute.
  * One overall goal/project can be: Create custom GPT which given a code base will prepare document, tests and review it.
- Enhancements to imbed:
  * Currently one needs to provide path where scraped URL content should be stored. This can be changed to default location
  * Langchain embedding is used. Check if it can be made generic to use other embedding offering APIs.

**Comment by thorwhalen** [comment]
## RAG-enabled custom GPTs for i2mint

Such GPTs will be able to help with the diagnosis and generation of documentation, tests and code (generation, review, refactor, etc.)

More specifically, these custom GPTs will be able to be used to 
* help contributors find contribution opportunities (documentation, tests, code)
  * where is documentation missing?
  * where are tests missing?
  * where should code be refactored?
  * what's the priority (dependency analysis, impact analysis...)
* generate code using `i2mint` tools and design philosophy
* help users find and understand these tools and their design

We'll use `i2mint` as our target knowledge base, but obviously we want to develop the tools so we can apply them to more general cases such as:
* one or a set of github organizations (containing code, discussions, wikis, issues...)
* any source of code, documentation, commentary, etc.
* any corpus (obviously, many of the tools developed will be able to be used in more general contexts)

### User stories and Use Cases

I can create tutorials on specific use cases that i2mint can help solving. 

I can get diagnoses on what documentation is missing, or weak, or possibly misaligned, and get help to create, edit and integrate this documentation. Documentation can come in the form of tutorials, doc strings, or comments. Tutorials can be generated from commented tests (see [test2doc](https://github.com/i2mint/test2doc)). 

I can get diagnoses on what tests are missing, or weak, and get help to create and integrate tests (doctests and pytest tests). 

I can get diagnoses on what code should be refactored, and help doing so. 

All my diagnoses can be enriched with a sense of priority, based on how the different functionalities depend on each other (the dependency analysis could be generated by an LLM, or made by an ast-aware python tool whose diagnosis is given to the LLM). 


### A Plan

Keep in mind that even if not explicitly stated, any concrete task is explicitly meant to make us move forward towards the functionalities mentioned above, but should be taken as an implicit excuse to develop (or improve) the more broadly reusable tools we need to complete the task, and once these reusable tools tasks are clear they should be made explicit. 

There's two main components to the custom GPT:
* the knowledge base (or "data")
* the prompt

We'll work on both, but the biggest work will probably be on the data. 

Here are a few verbs that describe what kind of operations we'll need to do on the data (and therefore build tools to do so effectively):
* Collect
* Prepare
* Search
* Inject

Note that web-services and UIs can be developed for any of these operations, but will especially be relevant for the search and inject parts. This is where the following can be useful (or should be learned and/or developed to become useful):
* JS-PY tools such as [jy](https://github.com/i2mint/jy) and [ju](https://github.com/i2mint/ju),
* visualization tools such as a [cosmograph](https://cosmograph.app/) widget, 
* webservice tools such as [fastAPI](https://fastapi.tiangolo.com/), [py2http](https://github.com/i2mint/py2http) and (in development) [qh](https://github.com/i2mint/qh), and 
* UI tools such as [rjsf](https://rjsf-team.github.io/react-jsonschema-form/docs/) (some tools in [ju](https://github.com/i2mint/ju)), [front](https://github.com/i2mint/front), [fastui](https://github.com/pydantic/FastUI) etc.

#### Collect

* clone repository (works for code only -- not issues, discussions or wikis)
  * don't need github API to clone
  * use [hubcap](https://github.com/thorwhalen/hubcap)'s `github_repo_text_aggregate` function
* use [hubcap](https://github.com/thorwhalen/hubcap) (has tools for code, but also issues, discussions and wikis), or directly the [github python API](https://github.com/PyGithub/PyGithub) (preferably, `hubcap`, so you can contribute towards that tool)
* scrape info (accumulate scraping tools in: [scraped](https://github.com/thorwhalen/scraped))

#### Prepare

By this I really mean organize and prepare the data so it can be accessed and processed. 
On one end of the spectrum, this could just mean aggregating text in one or several files. 
On the other end of the spectrum, this could mean segmenting, imbedding, and indexing the data, using a vector database etc.

In a first phase, where the outcome of our preparing phase will just be to aggregate the data into one or several files, our objective is (as always) to make this process as streamlined as possible. Our method to do so is (as always), to depend on abstract simple interfaces. For example, the core function that will create a text file by aggregating the content of several other local files will depend on an iterable of strings or a mapping (key-value where the value is a string). This function can then be reused in many situations, the only thing that is needed to adapt being then an adaptor that provides an interface to our data where it's at, from the format it's in. 

In a second phase, where we'll segment, imbed, and index our data, we want to again make sure we facade any component we use, favoring base interfaces. Work on this has been, and should continue being, accumulated in the [imbed](https://github.com/thorwhalen/imbed) project. 

So again: Use [base interfaces](https://docs.python.org/fr/3.10/library/collections.abc.html#collections-abstract-base-classes) (`Iterable`, `List`, `Mapping`, etc.) to facade your data access anytime you can.
This is where [dol](https://github.com/i2mint/dol) tools can help a lot. For example, using `dol.Files` with filters and key and value codecs.

#### Search

In a first phase we won't really have search functionality, since we'll be injecting our knowledge base via one or several aggregate files, and let the LLM decide what is relevant. 

In a second phase, which will be necessary as soon as there's too much data to fit in a few files, we'll have a vector DB that will contain indexed segments of our data. We'll then give the custom GPT a link to the vector DB, telling it to use it to search for relevant information. See this chatGPT conversation about [file constraints and vector DBs in custom GPTs](https://chatgpt.com/share/11f06a71-ab35-4448-8feb-3627c9c97197). 

In a third phase we can implement a search user interface to the vector DB so that the user can have some control over what they'll use. This can help a lot in mitigating problems of "basic RAG" (as described in the previous phase) because it enables the user (final user or developer) to diagnose the retrieval part of a RAG system and determine what needs to be done there (do we need to work on the segmentation, metadata, etc.?). 

#### Inject

For files, it's just uploading (or replacing) the files that are used by the knowledge base of the custom GPT. 

For vector DBs, it's specifying the link to it. 

If we want to implement a "controlled search" interface, we can 
* try getting that behavior via the prompt (try to use chatGPT as a search engine on the knowledge base, listing search results to the user and having them chose what gets used...
* implement a search and "search results set" functionality, along with a way to then inject the results into chatGPT (via the clipboard, or a chrome extension, or an atom desktop app (that uses openAI's API)).
